File: go/internal/app/alert_worker.go
package app

import (
	"context"
	"fmt"
	"log"
	"time"
)

type AlertWorker struct {
	sub          <-chan Event
	cancel       func()
	store        Store
	messageQueue *MessageQueue
}

func NewAlertWorker(pubsub *PubSub, store Store, buffer int, messageQueue *MessageQueue) *AlertWorker {
	sub, cancel := pubsub.Subscribe(buffer)
	return &AlertWorker{
		sub:          sub,
		cancel:       cancel,
		store:        store,
		messageQueue: messageQueue,
	}
}

func (w *AlertWorker) Run(ctx context.Context) {
	defer w.cancel()
	for {
		select {
		case <-ctx.Done():
			return
		case event, ok := <-w.sub:
			if !ok {
				return
			}
			w.handleEvent(ctx, event)
		}
	}
}

func (w *AlertWorker) handleEvent(ctx context.Context, event Event) {
	if event.Type != EventTypeVitalReceived {
		return
	}
	if !IsAbnormal(event.Vital) {
		return
	}

	reason := AlertReason(event.Vital)
	alert := Alert{
		VitalID:    event.Vital.ID,
		PatientID:  event.Vital.PatientID,
		Systolic:   event.Vital.Systolic,
		Diastolic:  event.Vital.Diastolic,
		TakenAt:    event.Vital.TakenAt,
		ReceivedAt: event.Vital.ReceivedAt,
		Reason:     reason,
		Status:     AlertStatusActive,
		Created:    time.Now().UTC(),
	}

	if _, err := w.store.AddAlert(ctx, alert); err != nil {
		log.Printf("alert worker failed to store alert: %v", err)
		return
	}

	log.Printf("[Alert] %s: %s", event.Vital.PatientID, reason)

	if w.messageQueue != nil {
		content := fmt.Sprintf("Alert: %s. Please retake your vitals.", reason)
		w.messageQueue.Enqueue(event.Vital.PatientID, content)
	}
}

File: node/src/domain/alertWorker.ts
import { MessageQueue } from "./messageQueue";
import {
  AlertStatusActive,
  Event,
  EventTypeVitalReceived,
  alertReason,
  isAbnormal,
} from "./models";
import { PubSub, Subscription } from "./pubsub";
import { Store } from "./store";

export class AlertWorker {
  private subscription: Subscription;
  private cancel: () => void;
  private messageQueue: MessageQueue | null;

  constructor(
    pubsub: PubSub,
    private store: Store,
    bufferSize: number,
    messageQueue?: MessageQueue
  ) {
    const { subscription, cancel } = pubsub.subscribe(bufferSize);
    this.subscription = subscription;
    this.cancel = cancel;
    this.messageQueue = messageQueue ?? null;
  }

  async run(signal?: AbortSignal): Promise<void> {
    const abortHandler = () => this.cancel();
    signal?.addEventListener("abort", abortHandler);

    try {
      for await (const event of this.subscription) {
        if (signal?.aborted) {
          break;
        }
        await this.handleEvent(event);
      }
    } finally {
      signal?.removeEventListener("abort", abortHandler);
      this.cancel();
    }
  }

  private async handleEvent(event: Event): Promise<void> {
    if (event.type !== EventTypeVitalReceived) {
      return;
    }
    if (!isAbnormal(event.vital)) {
      return;
    }

    const reason = alertReason(event.vital);

    try {
      await this.store.addAlert({
        vitalId: event.vital.id,
        patientId: event.vital.patientId,
        systolic: event.vital.systolic,
        diastolic: event.vital.diastolic,
        takenAt: event.vital.takenAt,
        receivedAt: event.vital.receivedAt,
        reason,
        status: AlertStatusActive,
        created: new Date(),
      });

      console.log(`[Alert] ${event.vital.patientId}: ${reason}`);

      if (this.messageQueue) {
        this.messageQueue.enqueue(
          event.vital.patientId,
          `Alert: ${reason}. Please retake your vitals.`
        );
      }
    } catch (error) {
      console.warn("alert worker failed to store alert", error);
    }
  }
}
